cylc/uiserver/handlers.py

```diff
@@ -13,109 +13,153 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-import json
-import os
 from asyncio import Queue
-import logging
+from functools import wraps
+import json
 import getpass
+import socket
+from typing import Callable, Union
 
 from graphene_tornado.tornado_graphql_handler import TornadoGraphQLHandler
 from graphql import get_default_backend
 from graphql_ws.constants import GRAPHQL_WS
-from jupyterhub import __version__ as jupyterhub_version
-from jupyterhub.services.auth import HubOAuthenticated
+from jupyter_server.base.handlers import JupyterHandler
 from tornado import web, websocket
 from tornado.ioloop import IOLoop
 
-from .websockets import authenticated as websockets_authenticated
-
-
-logger = logging.getLogger(__name__)
+from cylc.uiserver.websockets import authenticated as websockets_authenticated
 
 
 ME = getpass.getuser()
 
 
-def _authorised(req):
-    """Authorise a request.
-
-    Requires the request to be authenticated.
+def authorised(fun: Callable) -> Callable:
+    """Provides Cylc authorisation.
 
-    Currently returns True if the authenticated user is the same as the
-    user under which this UI Server is running.
-
-    Returns:
-        bool - True if the request passes authorisation, False if it fails.
+    When the UIS is run standalone (token-authenticated) application,
+    authorisation is deactivated, the bearer of the token has full privileges.
 
+    When the UIS is spawned by Jupyter Hub (hub authenticated), multi-user
+    access is permitted. Users are authorised by _authorise.
     """
-    user = req.get_current_user()
-    username = user.get('name', '?')
-    if username != ME:
-        logger.warning(f'Authorisation failed for {username}')
-        return False
-    return True
 
-
-def authorised(fun):
-    """Provides Cylc authorisation for multi-user setups."""
-
-    def _inner(self, *args, **kwargs):
+    @wraps(fun)
+    def _inner(
+        handler: 'CylcAppHandler',
+        *args,
+        **kwargs,
+    ):
         nonlocal fun
-        if not _authorised(self):
+        user: Union[
+            None,   # unauthenticated
+            bytes,  # token auth (bug in jupyter_server)
+            dict,   # hub auth
+            str,    # token auth
+
+        ] = handler.get_current_user()
+        if user is None or user == 'anonymous':
+            # user is not authenticated - additional protection incase the
+            # endpoint is not authenticate protected by mistake
+            # NOTE: Auth tests will hit this line unless mocked authentication
+            # is provided.
+            raise web.HTTPError(403, reason='Forbidden')
+        handler.log.critical(
+            f'authorise {user} '  # type: ignore
```

Comment:
Is it dangerous to log the User's token?

Reply:
Almost certainly, apologies this is some debug kicking around due to the problem of detecting token-authenticated sessions. Hopefully there will be a solution there in the coming days :crossed_fingers:.